### ISSUE TITLE
Convert Kubernetes events to traces

### ISSUE THREAD

**Issue by jpkrohling** [open-telemetry/opentelemetry-collector-contrib#7408]
PR #6885 added a Kubernetes events receiver, converting events into logs. It should also be possible to convert them into traces, as seen in the `kspan` project: https://github.com/weaveworks-experiments/kspan

cc @dmitryax, @hprajapati-splunk

**Comment by dmitryax** [comment]
It sounds interesting!

**Comment by TylerHelmuth** [comment]
I am very interested in this.  If there is still acceptance that this is a good idea I'd like to take a look.

**Comment by dmitryax** [comment]
It sounds reasonable to me. The thing is `k8sevents` receiver is going to be deprecated in favor of `k8sobjects` receiver. Not sure where this functionality should be added to. Maybe another component k8s trace receiver?

**Comment by jpkrohling** [comment]
There's a sudden interest in this :-) I know that @dgoscn is looking into this, and @yuriolisa has pinged me about something like this for the Operator. You all should talk :-)

**Comment by TylerHelmuth** [comment]
Ya @yuriolisa and I have talked briefly.

@dmitryax if the `k8sevents` receiver is deprecated we should definitely not add it there (and we should mark it's README accordingly).

I see 2 solutions:
  1. Create a new Receiver dedicated to converting k8s events to traces
  2. Update the `k8sevents` receiver to be able to be used in the traces pipeline where it can convert only k8s events to traces.

I like option 2 the most.  I am almost always in favor of less components and we already decided that k8sobjectsreceiver should replace k8seventsreceiver since k8sobjectsreceiver could also grab events. Adding a second component that interacts with events feels unnecessary.

The tricky part with option 2 is that adding the k8sobjectreciever to the traces pipeline only makes sense if it is configured to collect events.  At the moment I believe the configuration would need reworked some about to make it natural to configure it both for logs pipeline and traces pipeline.  I think a combination of good config interface and config validation will allow option 2 to succeed.

**Comment by dmitryax** [comment]
@TylerHelmuth That sounds good to me but it's unclear from your message which one you want to add this functionality to: `k8sevents` or `k8sobjects`. We can add it to `k8sevents` and deprecate only logs part of it

**Comment by TylerHelmuth** [comment]
I mistyped in my options, updated the comment to be correct.

I would prefer to add it to `k8sobjects` and deprecate `k8sevents` entirely.  All the functionality of the `k8sevents` component is met by the `k8sobjects` component, so we should continue to push that idea by adding new functionality to `k8sobjects`.

**Comment by dmitryax** [comment]
SGTM but please assess the current configuration interface of k8sobjects receiver. Ideally we want to have the same interface applicable to the both logs and traces

**Comment by TylerHelmuth** [comment]
Yup, I think proposing a config interface that make it usable for the traces pipeline will be the first step.

**Comment by yuriolisa** [comment]
@TylerHelmuth, @jpkrohling, thank you for approaching that. We can join the efforts to enable k8sobjects for the traces pipeline.

**Comment by TylerHelmuth** [comment]
Once we start implementing this feature, I propose we use https://github.com/honeycombio/kspan pending no response on https://github.com/weaveworks-experiments/kspan/issues/57.

**Comment by TylerHelmuth** [comment]
Config Proposal 1: we leave the config  alone and make the traces pipeline search for the `Events` object in the `objects` list.  If `Events` is present, make some traces.  If not, do nothing. Since `Validate` doesn't know which pipelines the component will be used for I'm not sure we could fail if the component is used in Traces without `Events`.
If we ever need some trace-specific event configuration that would be problematic.


Config Proposal 2: We update the config to have signal-specific object lists:

```
 k8sobjects:
    auth_type: serviceAccount
    trace_objects:
      - name: events
        mode: watch
        group: events.k8s.io
        namespaces: [default]
    log_objects:
      - name: pods
        mode: pull
        label_selector: environment in (production),tier in (frontend)
        field_selector: status.phase=Running
        interval: 15m
      - name: events
        mode: watch
        group: events.k8s.io
        namespaces: [default]
```

This option is more explicit and allows us to do validation that `trace_objects` contains `events. It also allows us to add trace-specific config if it comes up. Downside is that it is a breaking change and more verbose.

**Comment by jpkrohling** [comment]
I like the second proposal, we can even make it an acceptable change by moving the current `objects` to `log_objects` for a couple of versions before removing it.

**Comment by TylerHelmuth** [comment]
We have precedence doing this type of signal-specific config sections in components like the transformprocessor and filterprocessor, but do we have precedence in any Receivers?

**Comment by TylerHelmuth** [comment]
Looks like [sqlqueryreceiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/sqlqueryreceiver), [mongodbatlasreceiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/mongodbatlasreceiver) and [azureblobreceiver](https://github.com/open-telemetry/opentelemetry-collector-contrib/tree/main/receiver/azureblobreceiver) all have precedence for having log/trace specific configurations.

**Comment by TylerHelmuth** [comment]
I started looking at what it will take to use kspan as a library and it will require some refactoring to decouple exporting from event->span generation.  I'll start looking into this refactor, attempting to keep the internals of kspan untouched as much as possible.

At the moment kspan creates `tracesdk.SpanSnapshot` using a very old version of OTel Go SDK.  We'll need to do some dependency bumping, but also I think we'll need to start with the library returning us Go SDK spans and then convert them to pdata spans in the receiver.  I don't want to take on added pdata to kspan to start.

**Comment by TylerHelmuth** [comment]
I started digging into kspan internals to upgrade the Go SDK and while I have a solution, I really don't like it.  kspan expects the spans to be mutable and does a lot of modifications to span times and resources between "discovery" and exporting - OTel Spans are not mutable so I had to introduce an extra data structure. On top of that, the current project is not situated to be used as a library.  Using the existing code as a library would require lots of rework.

I am going to spend some time experimenting with moving the business logic into `k8sobjectsreceiver/internal`.  This will give us direct control over the heuristics, better testing opportunities, and allows us to use pdata which is the best data structure for kspan's internal cache (and prevents any extra data structure conversions).

If this experiment works out, we'll make sure the original kspan authors are credited in our internal implementation.

**Comment by yuriolisa** [comment]
@TylerHelmuth, do you have any updates regarding this topic?

**Comment by TylerHelmuth** [comment]
@yuriolisa after experimenting I have been unsuccessful on all fronts. 

Also, reviewing Honeycomb's internal use of kspan my conclusion is that using raw events provides the same experience. I am not claiming the logs and traces are the same, I believe traces are superior, but in this situation I wasn't gaining value from the span relationship that was being generated. We ended up using the spans themselves as a way to view an event, but the relationship between spans and their association in a trace wasn't helpful. When we started with kspan there wasn't a good way in OTel to get the events, but now there is. Ultimately I'd rather see k8s provide more native tracing, like it has already started doing, instead of spending time relating the events it generates as spans. 

All that to say, I spent a lot of time on this topic this past summer and my conclusion is that I'm no longer interested. We are now using the collector to gather events directly. Thanks for reminding me to come back to this issue and leave an update.
